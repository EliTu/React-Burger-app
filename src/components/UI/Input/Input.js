import React, { useRef, useEffect } from 'react';
import styles from './Input.module.css';
import PropTypes from 'prop-types';

const Input = ({
	elementType,
	elementConfig,
	value,
	validation,
	focused,
	handleChange,
	handleEnterPress,
	data,
}) => {
	// CSS Modules styles:
	const { Input, InvalidStyle, errorMessageStyle, ValidStyle } = styles;

	let inputElement = null;

	let errorMessageElement = (
		<p className={errorMessageStyle}>{validation.errorMessage}</p>
	);

	// Listen to keyboard enter click to submit form:
	const enterPressCallback = (data, event, func) => {
		if (event.key === 'Enter' && data !== 'search') func(event);
	};

	// Focus the first input field upon component mount
	const focusRef = useRef();
	useEffect(() => {
		if (focused) focusRef.current.focus();
	}, [focused]);

	// Set validation styles:
	let validationStyles = [];
	validationStyles =
		!validation.valid && validation.hasUserInput
			? InvalidStyle
			: validation.valid && validation.hasUserInput && value !== ''
			? ValidStyle
			: [];

	switch (elementType) {
		case 'input':
			inputElement = (
				<input
					ref={focusRef}
					className={validationStyles}
					{...elementConfig}
					value={value}
					onChange={handleChange}
					onKeyPress={event =>
						enterPressCallback(data, event, handleEnterPress)
					}
					data-test="input-test"
				/>
			);
			break;

		case 'textarea':
			inputElement = (
				<textarea
					{...elementConfig}
					value={value}
					onChange={handleChange}
					onKeyPress={event =>
						enterPressCallback(event, handleEnterPress)
					}
					data-test="textarea-test"
				/>
			);
			break;

		case 'select':
			inputElement = (
				<select
					value={value}
					onChange={handleChange}
					data-test="select-test"
				>
					{elementConfig.options.map(option => (
						<option
							value={option.value}
							key={option.value}
							disabled={option.disabled}
						>
							{option.displayValue}
						</option>
					))}
				</select>
			);
			break;

		default:
			inputElement = (
				<input
					{...elementConfig}
					value={value}
					className={!validation.valid ? InvalidStyle : null}
					onKeyPress={event =>
						enterPressCallback(event, handleEnterPress)
					}
				/>
			);
	}
	return (
		<div className={Input}>
			<label>{elementConfig.label}</label>
			{inputElement}
			{!validation.valid && validation.hasUserInput
				? errorMessageElement
				: null}
		</div>
	);
};

Input.propTypes = {
	elementType: PropTypes.string,
	elementConfig: PropTypes.object,
	validation: PropTypes.object,
	focused: PropTypes.bool,
	handleChange: PropTypes.func,
	handleEnterPress: PropTypes.func,
};

export default Input;
