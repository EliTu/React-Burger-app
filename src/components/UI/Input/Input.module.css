@import '../../../variables.css';

.Input {
    width: 100%;
    padding: 0.2rem;
    box-sizing: border-box;
    font-family: inherit;
}

.Input input,
.Input textarea,
.Input select {
    background-color: rgba(241, 238, 238, 0);
    color: var(--coloring-dark);
    font-size: 0.9rem;
    width: 60%;
    border: none;
    border-radius: 2px;
    border: 2px solid var(--coloring-opaque-dark);
    padding: 0.3rem 0.7rem;
    margin: 0.7rem auto;
    display: inline-block;
    transition: all 0.5s;
}

/* To disable up and down number spinners on number input field */
.Input input[type=number]::-webkit-inner-spin-button,
.Input input[type=number]::-webkit-outer-spin-button {
    -webkit-appearance: none;
    margin: 0;
}

.Input textarea {
    border: 2px solid var(--coloring-opaque-dark);
    overflow: auto;
    resize: none;
    text-align: start;
    margin: 0.5rem auto;
    height: 3rem;
    width: 75%;
    height: 5rem;
}

.Input label {
    width: 35%;
    font-weight: 700;
    display: inline-block;
    margin-bottom: 0.5rem;
    text-align: center;
}

.Input select option {
    text-align: center;
    background-color: rgba(241, 238, 238, 0.993);
    color: var(--coloring-dark);
}

.Input input:focus,
.Input select:focus {
    border: 2px solid var(--glow-light-blue);
}

.Input textarea:focus {
    border: 2px solid var(--glow-light-blue);
}

.Input .ValidStyle {
    border: 2px solid var(--glow-green-opaque-lighter);
}

.Input .InvalidStyle {
    border: 2px solid var(--invalid-red-dark);
}

.errorMessageStyle {
    margin: 0.1rem;

    font-size: 0.8rem;
    color: var(--invalid-red-light);
}

::placeholder {
    color: var(--coloring-opaque-dark);
}

/* Small screen */
@media screen and (max-width: 800px) {

    .Input input,
    .Input textarea,
    .Input select {
        font-size: 0.9rem;
        width: 70%;
        display: block;
    }

    .Input textarea {
        width: 85%;
        height: 5rem;

    }

    .Input label {
        width: 100%;
        display: block;
    }
}