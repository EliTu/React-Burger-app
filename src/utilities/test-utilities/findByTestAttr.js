// Used to find a specific DOM element that is specified with the 'data-test' attribute to better traverse while testing:
export const findByTestAttr = (component, attr) => {
	const wrapper = component.find(`[data-test='${attr}']`);
	return wrapper;
};
