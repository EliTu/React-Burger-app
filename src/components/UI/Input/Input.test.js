import React from 'react';
import Input from './Input';
import { shallow } from 'enzyme';
import { findByTestAttr } from '../../../utilities/test-utilities/findByTestAttr';

const onChange = jest.fn();
const onEnterPress = jest.fn();

const setComponentProps = (
	elemType,
	configType = 'text',
	value = 'abc',
	isValid = false,
	hasInput = true,
	options = []
) => {
	const props = {
		elementType: `${elemType}`,
		elementConfig: {
			type: `${configType}`,
			placeholder: 'abc',
			label: 'abc',
			options: [
				{
					value: 'fastest',
					displayValue: 'Fastest',
					key: '1',
				},
				{ value: 'cheapest', displayValue: 'Cheapest', key: '2' },
			],
		},
		value: `${value}`,
		validation: {
			required: true,
			hasUserInput: `${hasInput}`,
			valid: `${isValid}`,
			errorMessage: 'abc',
		},
		focused: true,
		handleChange: () => {},
		handleEnterPress: () => {},
		data: 'abc',
	};
	return props;
};

const setComponent = (props, changeFn = () => {}, enterFn = () => {}) => {
	const component = shallow(
		<Input {...props} handleChange={changeFn} handleEnterPress={enterFn} />
	);
	return component;
};

describe('<Input/>', () => {
	describe('Component (not testing the input fields)', () => {
		it('should render an Input component withot errors', () => {
			let props = setComponentProps('input', 'text', 'abc', true);
			let component = setComponent(props);
			expect(component).toMatchSnapshot();
			expect(component).toBeTruthy();
			expect(component.hasClass('Input')).toBe(true);
			expect(component.length).toBe(1);
			expect(component.length).not.toBe(2);
		});
	});

	describe('Type: input="text" test', () => {
		let props = setComponentProps('input');
		let component = setComponent(props);
		let input = findByTestAttr(component, 'input-test');

		it('should render the component without errors', () => {
			expect(component).toMatchSnapshot();
			expect(input.length).toBe(1);
		});

		it('should not render if elementType is not input ', () => {
			props = setComponentProps('password');
			component = setComponent(props);
			input = findByTestAttr(component, 'input-test');

			expect(component).toMatchSnapshot();
			expect(input.length).toBe(0);
		});

		it('should call the onChange event callback function on user input', () => {
			props = setComponentProps('input', null, '');
			component = setComponent(props, onChange);
			input = findByTestAttr(component, 'input-test');

			input.simulate('change', {
				target: { value: 'abc' },
			});

			expect(onChange).toHaveBeenCalled();
		});

		it('should call the onKeyPress event callback function on user press enter', () => {
			props = setComponentProps('input');
			component = setComponent(props, null, onEnterPress);
			input = findByTestAttr(component, 'input-test');

			input.simulate('keypress', { key: 'Enter' });

			expect(onEnterPress).toHaveBeenCalled();
		});
	});

	describe('Type: input="textarea" test', () => {
		let props = setComponentProps('textarea', '');
		let component = setComponent(props);
		let textarea = findByTestAttr(component, 'textarea-test');
		it('should render without errors', () => {
			expect(component).toMatchSnapshot();
			expect(textarea.length).toBe(1);
			expect(textarea.length).not.toBe(2);
		});

		it('should call the onKeyPress event callback function on user press enter', () => {
			component = setComponent(props, null, onEnterPress);

			textarea.simulate('keypress', { key: 'Enter' });

			expect(onEnterPress).toHaveBeenCalled();
		});
	});

	describe('Type: input="select" test', () => {
		let props = setComponentProps('select');
		let component = setComponent(props);
		let select = findByTestAttr(component, 'select-test');

		it('should render without errors', () => {
			expect(component).toMatchSnapshot();
			expect(select.length).toBe(1);
			expect(select.find('option').length).toBe(2);
		});

		it('should call the onChange event callback function on user change option', () => {
			select.simulate('change', { target: { value: 'efg' } });
			expect(onChange).toHaveBeenCalled();
		});
	});
});
