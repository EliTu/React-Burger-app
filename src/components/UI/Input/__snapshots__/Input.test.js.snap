// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Input/> Component (not testing the input fields) should render an Input component withot errors 1`] = `
<div
  className="Input"
>
  <label>
    abc
  </label>
  <input
    className="ValidStyle"
    data-test="input-test"
    label="abc"
    onChange={[Function]}
    onKeyPress={[Function]}
    options={
      Array [
        Object {
          "displayValue": "Fastest",
          "key": "1",
          "value": "fastest",
        },
        Object {
          "displayValue": "Cheapest",
          "key": "2",
          "value": "cheapest",
        },
      ]
    }
    placeholder="abc"
    type="text"
    value="abc"
  />
</div>
`;

exports[`<Input/> Type: input="select" test should render without errors 1`] = `
<div
  className="Input"
>
  <label>
    abc
  </label>
  <select
    data-test="select-test"
    onChange={[Function]}
    value="abc"
  >
    <option
      key="fastest"
      value="fastest"
    >
      Fastest
    </option>
    <option
      key="cheapest"
      value="cheapest"
    >
      Cheapest
    </option>
  </select>
</div>
`;

exports[`<Input/> Type: input="text" test should not render if elementType is not input  1`] = `
<div
  className="Input"
>
  <label>
    abc
  </label>
  <input
    className={null}
    label="abc"
    onKeyPress={[Function]}
    options={
      Array [
        Object {
          "displayValue": "Fastest",
          "key": "1",
          "value": "fastest",
        },
        Object {
          "displayValue": "Cheapest",
          "key": "2",
          "value": "cheapest",
        },
      ]
    }
    placeholder="abc"
    type="text"
    value="abc"
  />
</div>
`;

exports[`<Input/> Type: input="text" test should render the component without errors 1`] = `
<div
  className="Input"
>
  <label>
    abc
  </label>
  <input
    className="ValidStyle"
    data-test="input-test"
    label="abc"
    onChange={[Function]}
    onKeyPress={[Function]}
    options={
      Array [
        Object {
          "displayValue": "Fastest",
          "key": "1",
          "value": "fastest",
        },
        Object {
          "displayValue": "Cheapest",
          "key": "2",
          "value": "cheapest",
        },
      ]
    }
    placeholder="abc"
    type="text"
    value="abc"
  />
</div>
`;

exports[`<Input/> Type: input="textarea" test should render without errors 1`] = `
<div
  className="Input"
>
  <label>
    abc
  </label>
  <textarea
    data-test="textarea-test"
    label="abc"
    onChange={[Function]}
    onKeyPress={[Function]}
    options={
      Array [
        Object {
          "displayValue": "Fastest",
          "key": "1",
          "value": "fastest",
        },
        Object {
          "displayValue": "Cheapest",
          "key": "2",
          "value": "cheapest",
        },
      ]
    }
    placeholder="abc"
    type=""
    value="abc"
  />
</div>
`;
